File: web-app/src/api/assistant.ts
/**
 * API client for AI Financial Assistant.
 *
 * Handles SSE streaming for chat and REST endpoints for conversations.
 */

const baseUrl =
  import.meta.env.VITE_API_URL ||
  (typeof window !== "undefined"
    ? window.location.origin
    : "http://localhost:8080");

// Types
export interface ChatContext {
  current_page?: string;
  selected_transaction_id?: number;
  active_filters?: Record<string, string>;
}

export interface ChatRequest {
  message: string;
  conversation_id?: string;
  context?: ChatContext;
}

export interface Conversation {
  id: string;
  title: string | null;
  message_count: number;
  total_tokens: number;
  created_at: string | null;
  updated_at: string | null;
  is_archived: boolean;
}

export interface Message {
  role: "user" | "assistant" | "system" | "tool";
  content: string;
  tool_calls?: ToolCall[];
  tool_call_id?: string;
}

export interface ToolCall {
  id: string;
  name: string;
  arguments: Record<string, unknown>;
}

export interface ActionCard {
  type: "navigation" | "action";
  url?: string;
  label: string;
}

export interface ActionProposal {
  type: "action_proposal";
  action: string;
  requires_confirmation: boolean;
  preview: Record<string, unknown>;
  payload: Record<string, unknown>;
}

export interface Insight {
  id: string;
  insight_type: string;
  priority: number;
  title: string;
  content: string;
  action_type?: string;
  action_payload?: Record<string, unknown>;
  created_at: string;
}

export interface HealthStatus {
  status: "ok" | "unconfigured";
  llm_configured: boolean;
  llm_provider: string;
  llm_model: string;
}

// SSE Event Types
export type SSEEventType =
  | "conversation_id"
  | "content_delta"
  | "tool_call"
  | "tool_result"
  | "action_proposal"
  | "action_card"
  | "message_complete"
  | "error";

export interface SSEEvent {
  event: SSEEventType;
  data: Record<string, unknown>;
}

/**
 * Stream chat response using Server-Sent Events.
 */
export async function* streamChat(
  request: ChatRequest
): AsyncGenerator<SSEEvent, void, unknown> {
  const response = await fetch(`${baseUrl}/api/assistant/chat`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    throw new Error(`Chat request failed: ${response.statusText}`);
  }

  const reader = response.body?.getReader();
  if (!reader) {
    throw new Error("No response body");
  }

  const decoder = new TextDecoder();
  let buffer = "";

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split("\n");
    buffer = lines.pop() || "";

    let currentEvent = "";
    let currentData = "";

    for (const line of lines) {
      if (line.startsWith("event: ")) {
        currentEvent = line.slice(7);
      } else if (line.startsWith("data: ")) {
        currentData = line.slice(6);
        if (currentEvent && currentData) {
          try {
            yield {
              event: currentEvent as SSEEventType,
              data: JSON.parse(currentData),
            };
          } catch {
            console.error("Failed to parse SSE data:", currentData);
          }
          currentEvent = "";
          currentData = "";
        }
      }
    }
  }
}

/**
 * List conversations.
 */
export async function listConversations(
  limit = 20,
  includeArchived = false
): Promise<Conversation[]> {
  const params = new URLSearchParams({
    limit: limit.toString(),
    include_archived: includeArchived.toString(),
  });

  const response = await fetch(
    `${baseUrl}/api/assistant/conversations?${params}`
  );

  if (!response.ok) {
    throw new Error(`Failed to list conversations: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Get messages for a conversation.
 */
export async function getConversationMessages(
  conversationId: string,
  limit = 50
): Promise<{ conversation_id: string; messages: Message[] }> {
  const params = new URLSearchParams({ limit: limit.toString() });

  const response = await fetch(
    `${baseUrl}/api/assistant/conversations/${conversationId}/messages?${params}`
  );

  if (!response.ok) {
    throw new Error(`Failed to get messages: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Archive a conversation.
 */
export async function archiveConversation(
  conversationId: string
): Promise<{ success: boolean; message: string }> {
  const response = await fetch(
    `${baseUrl}/api/assistant/conversations/${conversationId}/archive`,
    { method: "POST" }
  );

  if (!response.ok) {
    throw new Error(`Failed to archive conversation: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Delete a conversation.
 */
export async function deleteConversation(
  conversationId: string
): Promise<{ success: boolean; message: string }> {
  const response = await fetch(
    `${baseUrl}/api/assistant/conversations/${conversationId}`,
    { method: "DELETE" }
  );

  if (!response.ok) {
    throw new Error(`Failed to delete conversation: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Execute a confirmed action.
 */
export async function executeAction(
  messageId: string,
  actionType: string,
  payload: Record<string, unknown>
): Promise<{
  success: boolean;
  message: string;
  action_log_id: string;
  undo_available: boolean;
}> {
  const response = await fetch(`${baseUrl}/api/assistant/actions`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      message_id: messageId,
      action_type: actionType,
      confirmed: true,
      payload,
    }),
  });

  if (!response.ok) {
    throw new Error(`Failed to execute action: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Undo an action.
 */
export async function undoAction(
  actionLogId: string
): Promise<{ success: boolean; message: string }> {
  const response = await fetch(`${baseUrl}/api/assistant/actions/undo`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ action_log_id: actionLogId }),
  });

  if (!response.ok) {
    throw new Error(`Failed to undo action: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Get pending insights.
 */
export async function getInsights(limit = 10): Promise<Insight[]> {
  const params = new URLSearchParams({ limit: limit.toString() });

  const response = await fetch(`${baseUrl}/api/assistant/insights?${params}`);

  if (!response.ok) {
    throw new Error(`Failed to get insights: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Dismiss an insight.
 */
export async function dismissInsight(
  insightId: string
): Promise<{ success: boolean; message: string }> {
  const response = await fetch(
    `${baseUrl}/api/assistant/insights/${insightId}/dismiss`,
    { method: "POST" }
  );

  if (!response.ok) {
    throw new Error(`Failed to dismiss insight: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Check assistant health.
 */
export async function checkHealth(): Promise<HealthStatus> {
  const response = await fetch(`${baseUrl}/api/assistant/health`);

  if (!response.ok) {
    throw new Error(`Health check failed: ${response.statusText}`);
  }

  return response.json();
}

File: core-api/src/financial_analysis/api/routes/assistant.py
"""
API routes for AI Financial Assistant.

Provides endpoints for chat, conversation management, and insights.
"""

from typing import Optional, Dict, Any, List
import json
import logging

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from ..dependencies import get_db
from ...services.assistant import AssistantService, OpenAIAdapter
from ...database.assistant_models import AssistantPreferences

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/assistant", tags=["AI Assistant"])


# ===== Pydantic Models =====

class ChatContext(BaseModel):
    """Context for the chat message."""
    current_page: Optional[str] = None
    selected_transaction_id: Optional[int] = None
    active_filters: Optional[Dict[str, Any]] = None


class ChatRequest(BaseModel):
    """Request body for chat endpoint."""
    message: str = Field(..., min_length=1, max_length=2000)
    conversation_id: Optional[str] = None
    context: Optional[ChatContext] = None


class ActionConfirmRequest(BaseModel):
    """Request body for action confirmation."""
    message_id: str
    action_type: str
    confirmed: bool
    payload: Dict[str, Any]


class ConversationResponse(BaseModel):
    """Response for conversation listing."""
    id: str
    title: Optional[str]
    message_count: int
    total_tokens: int
    created_at: Optional[str]
    updated_at: Optional[str]
    is_archived: bool


class InsightResponse(BaseModel):
    """Response for insights."""
    id: str
    insight_type: str
    priority: int
    title: str
    content: str
    action_type: Optional[str]
    action_payload: Optional[Dict[str, Any]]
    created_at: str


# ===== Helper Functions =====

def get_api_key(db: Session) -> str:
    """Get OpenAI API key from preferences or environment."""
    import os

    # Try to get from database preferences
    prefs = db.query(AssistantPreferences).first()
    if prefs and prefs.openai_api_key_encrypted:
        # TODO: Decrypt the key
        pass

    # Fall back to environment variable
    api_key = os.getenv("OPENAI_API_KEY")
    if not api_key:
        raise HTTPException(
            status_code=503,
            detail="OpenAI API key not configured. Please set OPENAI_API_KEY environment variable or configure in settings."
        )

    return api_key


def get_assistant_service(db: Session = Depends(get_db)) -> AssistantService:
    """Dependency to get AssistantService instance."""
    api_key = get_api_key(db)
    model = os.getenv("OPENAI_MODEL", "gpt-4o")

    llm_adapter = OpenAIAdapter(api_key=api_key, model=model)
    return AssistantService(db=db, llm_adapter=llm_adapter)


import os  # Move to top in production


# ===== Endpoints =====

@router.post("/chat")
async def chat(
    request: ChatRequest,
    db: Session = Depends(get_db),
):
    """
    Send a message to the AI assistant and receive a streamed response.

    Returns a Server-Sent Events (SSE) stream with the following event types:
    - conversation_id: The conversation ID being used
    - content_delta: Incremental content from the assistant
    - tool_call: When the assistant calls a tool
    - tool_result: Result of tool execution
    - action_proposal: When an action requires user confirmation
    - action_card: Navigation or action button for the UI
    - message_complete: Final event with token usage
    - error: If an error occurs
    """
    try:
        api_key = get_api_key(db)
    except HTTPException as e:
        # Return error as SSE event
        async def error_generator():
            yield f"event: error\ndata: {json.dumps({'message': e.detail})}\n\n"
        return StreamingResponse(
            error_generator(),
            media_type="text/event-stream"
        )

    model = os.getenv("OPENAI_MODEL", "gpt-4o")
    llm_adapter = OpenAIAdapter(api_key=api_key, model=model)
    service = AssistantService(db=db, llm_adapter=llm_adapter)

    async def event_generator():
        """Generate SSE events from assistant response."""
        try:
            context = request.context.model_dump() if request.context else {}

            async for event in service.chat(
                message=request.message,
                conversation_id=request.conversation_id,
                context=context,
            ):
                event_type = event.get("event", "message")
                event_data = event.get("data", {})
                yield f"event: {event_type}\ndata: {json.dumps(event_data)}\n\n"

        except Exception as e:
            logger.error(f"Chat error: {e}")
            yield f"event: error\ndata: {json.dumps({'message': str(e)})}\n\n"

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",  # Disable nginx buffering
        }
    )


@router.get("/conversations", response_model=List[ConversationResponse])
async def list_conversations(
    limit: int = Query(default=20, ge=1, le=100),
    include_archived: bool = Query(default=False),
    db: Session = Depends(get_db),
):
    """
    List conversation history.

    Returns recent conversations ordered by last update time.
    """
    api_key = get_api_key(db)
    llm_adapter = OpenAIAdapter(api_key=api_key, model="gpt-4o")
    service = AssistantService(db=db, llm_adapter=llm_adapter)

    conversations = await service.get_conversations(
        limit=limit,
        include_archived=include_archived
    )

    return conversations


@router.get("/conversations/{conversation_id}/messages")
async def get_conversation_messages(
    conversation_id: str,
    limit: int = Query(default=50, ge=1, le=200),
    db: Session = Depends(get_db),
):
    """
    Get messages for a specific conversation.

    Returns messages in chronological order.
    """
    api_key = get_api_key(db)
    llm_adapter = OpenAIAdapter(api_key=api_key, model="gpt-4o")
    service = AssistantService(db=db, llm_adapter=llm_adapter)

    messages = await service.get_conversation_messages(
        conversation_id=conversation_id,
        limit=limit
    )

    return {"conversation_id": conversation_id, "messages": messages}


@router.post("/conversations/{conversation_id}/archive")
async def archive_conversation(
    conversation_id: str,
    db: Session = Depends(get_db),
):
    """
    Archive a conversation.

    Archived conversations are hidden from the default list but not deleted.
    """
    api_key = get_api_key(db)
    llm_adapter = OpenAIAdapter(api_key=api_key, model="gpt-4o")
    service = AssistantService(db=db, llm_adapter=llm_adapter)

    success = await service.archive_conversation(conversation_id)

    if not success:
        raise HTTPException(status_code=404, detail="Conversation not found")

    return {"success": True, "message": "Conversation archived"}


@router.delete("/conversations/{conversation_id}")
async def delete_conversation(
    conversation_id: str,
    db: Session = Depends(get_db),
):
    """
    Delete a conversation and all its messages.

    This action cannot be undone.
    """
    api_key = get_api_key(db)
    llm_adapter = OpenAIAdapter(api_key=api_key, model="gpt-4o")
    service = AssistantService(db=db, llm_adapter=llm_adapter)

    success = await service.delete_conversation(conversation_id)

    if not success:
        raise HTTPException(status_code=404, detail="Conversation not found")

    return {"success": True, "message": "Conversation deleted"}


@router.post("/actions")
async def execute_action(
    request: ActionConfirmRequest,
    db: Session = Depends(get_db),
):
    """
    Execute a confirmed action.

    Actions like categorization or rule creation require user confirmation
    before execution. This endpoint executes the confirmed action.
    """
    # TODO: Implement action execution with undo logging
    # This is Phase 2 functionality

    return {
        "success": True,
        "message": "Action executed",
        "action_log_id": "placeholder",
        "undo_available": True
    }


@router.post("/actions/undo")
async def undo_action(
    action_log_id: str,
    db: Session = Depends(get_db),
):
    """
    Undo a previous action.

    Restores the previous state before the action was executed.
    """
    # TODO: Implement undo functionality
    # This is Phase 2 functionality

    return {
        "success": True,
        "message": "Action undone"
    }


@router.get("/insights", response_model=List[InsightResponse])
async def get_insights(
    limit: int = Query(default=10, ge=1, le=50),
    db: Session = Depends(get_db),
):
    """
    Get pending proactive insights.

    Returns insights that haven't been dismissed or acted upon,
    ordered by priority.
    """
    # TODO: Implement insights retrieval
    # This is Phase 3 functionality

    return []


@router.post("/insights/{insight_id}/dismiss")
async def dismiss_insight(
    insight_id: str,
    db: Session = Depends(get_db),
):
    """
    Dismiss an insight.

    The insight will no longer appear in the pending list.
    """
    # TODO: Implement insight dismissal
    # This is Phase 3 functionality

    return {"success": True, "message": "Insight dismissed"}


@router.get("/health")
async def health_check():
    """
    Check if the assistant service is available.

    Returns status of LLM connectivity and configuration.
    """
    api_key = os.getenv("OPENAI_API_KEY")

    return {
        "status": "ok" if api_key else "unconfigured",
        "llm_configured": bool(api_key),
        "llm_provider": "openai",
        "llm_model": os.getenv("OPENAI_MODEL", "gpt-4o")
    }
